File: cpp/src/storage/manifest.cpp
// Copyright 2023 Zilliz
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "storage/manifest.h"
#include <algorithm>
#include <memory>
#include "arrow/filesystem/filesystem.h"

namespace milvus_storage {

Manifest::Manifest(std::shared_ptr<Schema> schema) : schema_(std::move(schema)) {}

const std::shared_ptr<Schema> Manifest::schema() { return schema_; }

void Manifest::add_scalar_fragment(Fragment&& fragment) { scalar_fragments_.push_back(fragment); }

void Manifest::add_vector_fragment(Fragment&& fragment) { vector_fragments_.push_back(fragment); }

void Manifest::add_delete_fragment(Fragment&& fragment) { delete_fragments_.push_back(fragment); }

void Manifest::add_blob(Blob&& blob) { blobs_.emplace_back(blob); }

const FragmentVector& Manifest::scalar_fragments() const { return scalar_fragments_; }

const FragmentVector& Manifest::vector_fragments() const { return vector_fragments_; }

const FragmentVector& Manifest::delete_fragments() const { return delete_fragments_; }

bool Manifest::has_blob(const std::string& name) {
  auto iter = std::find_if(blobs_.begin(), blobs_.end(), [&](Blob& blob) { return blob.name == name; });
  return iter != blobs_.end();
}

void Manifest::remove_blob_if_exist(const std::string& name) {
  std::remove_if(blobs_.begin(), blobs_.end(), [&](Blob& blob) { return blob.name == name; });
}

Result<Blob> Manifest::get_blob(const std::string& name) {
  auto iter = std::find_if(blobs_.begin(), blobs_.end(), [&](Blob& blob) { return blob.name == name; });
  if (iter == blobs_.end()) {
    return Status::FileNotFound("blob not found");
  }
  return *iter;
}

const std::vector<Blob>& Manifest::blobs() const { return blobs_; }

int64_t Manifest::version() const { return version_; }

void Manifest::set_version(int64_t version) { version_ = version; }

Result<manifest_proto::Manifest> Manifest::ToProtobuf() const {
  manifest_proto::Manifest manifest;
  manifest.set_version(version_);
  for (auto& fragment : vector_fragments_) {
    manifest.mutable_vector_fragments()->AddAllocated(fragment.ToProtobuf().release());
  }
  for (auto& fragment : scalar_fragments_) {
    manifest.mutable_scalar_fragments()->AddAllocated(fragment.ToProtobuf().release());
  }
  for (auto& fragment : delete_fragments_) {
    manifest.mutable_delete_fragments()->AddAllocated(fragment.ToProtobuf().release());
  }
  for (auto& blob : blobs_) {
    manifest.mutable_blobs()->AddAllocated(blob.ToProtobuf().release());
  }

  ASSIGN_OR_RETURN_NOT_OK(auto schema_proto, schema_->ToProtobuf());
  manifest.set_allocated_schema(schema_proto.release());
  return manifest;
}

void Manifest::FromProtobuf(const manifest_proto::Manifest& manifest) {
  schema_ = std::make_shared<Schema>();
  schema_->FromProtobuf(manifest.schema());

  for (auto& fragment : manifest.vector_fragments()) {
    vector_fragments_.emplace_back(*Fragment::FromProtobuf(fragment).release());
  }

  for (auto& fragment : manifest.scalar_fragments()) {
    scalar_fragments_.emplace_back(*Fragment::FromProtobuf(fragment).release());
  }

  for (auto& fragment : manifest.delete_fragments()) {
    delete_fragments_.emplace_back(*Fragment::FromProtobuf(fragment).release());
  }

  for (auto& blob : manifest.blobs()) {
    blobs_.emplace_back(Blob::FromProtobuf(blob));
  }

  version_ = manifest.version();
}

Status Manifest::WriteManifestFile(const Manifest* manifest, arrow::io::OutputStream* output) {
  ASSIGN_OR_RETURN_NOT_OK(auto proto_manifest, manifest->ToProtobuf());
  auto size = proto_manifest.ByteSizeLong();
  char* buffer = new char[size];
  proto_manifest.SerializeToArray(buffer, static_cast<int>(size));
  RETURN_ARROW_NOT_OK(output->Write(buffer, size));
  delete[] buffer;
  return Status::OK();
}

Result<std::shared_ptr<Manifest>> Manifest::ParseFromFile(std::shared_ptr<arrow::io::InputStream> istream,
                                                          arrow::fs::FileInfo& file_info) {
  auto size = file_info.size();
  char* buffer = new char[size];
  auto res = istream->Read(size, buffer);
  if (!res.ok()) {
    delete[] buffer;
    return Status::ArrowError(res.status().ToString());
  }

  manifest_proto::Manifest proto_manifest;
  proto_manifest.ParseFromArray(buffer, size);
  auto manifest = std::make_shared<Manifest>();
  manifest->FromProtobuf(proto_manifest);
  delete[] buffer;
  return manifest;
}
}  // namespace milvus_storage

File: go/storage/manifest/manifest.go
package manifest

import (
	"fmt"

	"github.com/apache/arrow/go/v12/arrow"
	"github.com/golang/protobuf/proto"
	"github.com/milvus-io/milvus-storage/go/common/log"
	"github.com/milvus-io/milvus-storage/go/file/blob"
	"github.com/milvus-io/milvus-storage/go/file/fragment"
	"github.com/milvus-io/milvus-storage/go/io/fs"
	"github.com/milvus-io/milvus-storage/go/io/fs/file"
	"github.com/milvus-io/milvus-storage/go/proto/manifest_proto"
	"github.com/milvus-io/milvus-storage/go/storage/schema"
)

type Manifest struct {
	schema          *schema.Schema
	ScalarFragments fragment.FragmentVector
	vectorFragments fragment.FragmentVector
	deleteFragments fragment.FragmentVector
	blobs           []blob.Blob
	version         int64
}

func NewManifest(schema *schema.Schema) *Manifest {
	return &Manifest{
		schema: schema,
	}
}

func Init() *Manifest {
	return &Manifest{
		schema: schema.NewSchema(arrow.NewSchema(nil, nil), schema.DefaultSchemaOptions()),
	}
}

func (m *Manifest) Copy() *Manifest {
	copied := *m
	return &copied
}

func (m *Manifest) GetSchema() *schema.Schema {
	return m.schema
}

func (m *Manifest) AddScalarFragment(fragment fragment.Fragment) {
	m.ScalarFragments = append(m.ScalarFragments, fragment)
}

func (m *Manifest) AddVectorFragment(fragment fragment.Fragment) {
	m.vectorFragments = append(m.vectorFragments, fragment)
}

func (m *Manifest) AddDeleteFragment(fragment fragment.Fragment) {
	m.deleteFragments = append(m.deleteFragments, fragment)
}

func (m *Manifest) GetScalarFragments() fragment.FragmentVector {
	return m.ScalarFragments
}

func (m *Manifest) GetVectorFragments() fragment.FragmentVector {
	return m.vectorFragments
}

func (m *Manifest) GetDeleteFragments() fragment.FragmentVector {
	return m.deleteFragments
}

func (m *Manifest) Version() int64 {
	return m.version
}

func (m *Manifest) SetVersion(version int64) {
	m.version = version
}

func (m *Manifest) ToProtobuf() (*manifest_proto.Manifest, error) {
	manifest := &manifest_proto.Manifest{}
	manifest.Version = m.version
	for _, vectorFragment := range m.vectorFragments {
		manifest.VectorFragments = append(manifest.VectorFragments, vectorFragment.ToProtobuf())
	}
	for _, scalarFragment := range m.ScalarFragments {
		manifest.ScalarFragments = append(manifest.ScalarFragments, scalarFragment.ToProtobuf())
	}
	for _, deleteFragment := range m.deleteFragments {
		manifest.DeleteFragments = append(manifest.DeleteFragments, deleteFragment.ToProtobuf())
	}

	for _, blob := range m.blobs {
		manifest.Blobs = append(manifest.Blobs, blob.ToProtobuf())
	}

	schemaProto, err := m.schema.ToProtobuf()
	if err != nil {
		return nil, err
	}
	manifest.Schema = schemaProto

	return manifest, nil
}

func (m *Manifest) FromProtobuf(manifest *manifest_proto.Manifest) error {
	err := m.schema.FromProtobuf(manifest.Schema)
	if err != nil {
		return err
	}

	for _, vectorFragment := range manifest.VectorFragments {
		m.vectorFragments = append(m.vectorFragments, fragment.FromProtobuf(vectorFragment))
	}

	for _, scalarFragment := range manifest.ScalarFragments {
		m.ScalarFragments = append(m.ScalarFragments, fragment.FromProtobuf(scalarFragment))
	}

	for _, deleteFragment := range manifest.DeleteFragments {
		m.deleteFragments = append(m.deleteFragments, fragment.FromProtobuf(deleteFragment))
	}

	for _, b := range manifest.Blobs {
		m.blobs = append(m.blobs, blob.FromProtobuf(b))
	}

	m.version = manifest.Version
	return nil
}

func WriteManifestFile(manifest *Manifest, output file.File) error {
	protoManifest, err := manifest.ToProtobuf()
	if err != nil {
		return err
	}

	bytes, err := proto.Marshal(protoManifest)
	if err != nil {
		return fmt.Errorf("write manifest file: %w", err)
	}
	write, err := output.Write(bytes)
	if err != nil {
		return fmt.Errorf("write manifest file: %w", err)
	}
	if write != len(bytes) {
		return fmt.Errorf("failed to write whole file, expect: %v, actual: %v", len(bytes), write)
	}
	if err = output.Close(); err != nil {
		return err
	}
	return nil
}

func (m *Manifest) HasBlob(name string) bool {
	for _, b := range m.blobs {
		if b.Name == name {
			return true
		}
	}

	return false
}

func (m *Manifest) AddBlob(blob blob.Blob) {
	m.blobs = append(m.blobs, blob)
}

func (m *Manifest) RemoveBlobIfExist(name string) {
	idx := -1
	for i, b := range m.blobs {
		if b.Name == name {
			idx = i
			break
		}
	}

	m.blobs = append(m.blobs[0:idx], m.blobs[idx+1:]...)
}

func (m *Manifest) GetBlob(name string) (blob.Blob, bool) {
	for _, b := range m.blobs {
		if b.Name == name {
			return b, true
		}
	}

	return blob.Blob{}, false
}

func ParseFromFile(f fs.Fs, path string) (*Manifest, error) {
	manifest := Init()
	manifestProto := &manifest_proto.Manifest{}

	buf, err := f.ReadFile(path)
	if err != nil {
		return nil, err
	}
	err = proto.Unmarshal(buf, manifestProto)
	if err != nil {
		log.Error("Failed to unmarshal manifest proto", log.String("err", err.Error()))
		return nil, fmt.Errorf("parse from file: %w", err)
	}
	manifest.FromProtobuf(manifestProto)

	return manifest, nil
}

// TODO REMOVE BELOW CODE

type DataFile struct {
	path string
	cols []string
}

func (d *DataFile) Path() string {
	return d.path
}

func NewDataFile(path string) *DataFile {
	return &DataFile{path: path}
}
